crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,767 @@
+        return 0;
+    }
+
+    if ((pp = OSSL_CRMF_POPO_new()) == NULL)
+        goto oom;
+    pp->type = ppmtd;
+
+    switch (ppmtd) {
+    case OSSL_CRMF_POPO_NONE:
+        OSSL_CRMF_POPO_free(pp);
+        OSSL_CRMF_POPO_free(crm->popo);
+        crm->popo = NULL;
+        return 1;
+    case OSSL_CRMF_POPO_RAVERIFIED:
+        if ((pp->value.raVerified = ASN1_NULL_new()) == NULL)
+            goto oom;
+        break;
+
+    case OSSL_CRMF_POPO_SIGNATURE:
+        {
+            OSSL_CRMF_POPOSIGNINGKEY *ps = OSSL_CRMF_POPOSIGNINGKEY_new();
+            if (ps == NULL
+                || !CRMF_poposigningkey_init(ps, crm->certReq, pkey, dgst)) {
+                OSSL_CRMF_POPOSIGNINGKEY_free(ps);
+                goto err;
+            }
+            pp->value.signature = ps;
+        }
+        break;
+
+    case OSSL_CRMF_POPO_KEYENC:
+        if ((pp->value.keyEncipherment = OSSL_CRMF_POPOPRIVKEY_new()) == NULL)
+            goto oom;
+        pp->value.keyEncipherment->type = OSSL_CRMF_POPOPRIVKEY_SUBSEQUENTMESSAGE;
+        if ((pp->value.keyEncipherment->value.subsequentMessage =
+             ASN1_INTEGER_new()) == NULL
+            ||
+            !ASN1_INTEGER_set(pp->value.keyEncipherment->value.subsequentMessage,
+                              OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT))
+            goto oom;
+        break;
+
+    default:
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO,
+                CRMF_R_UNSUPPORTED_METHOD_FOR_CREATING_POPO);
+        goto err;
+    }
+
+    OSSL_CRMF_POPO_free(crm->popo);
+    crm->popo = pp;
+
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO, ERR_R_MALLOC_FAILURE);
+ err:
+    OSSL_CRMF_POPO_free(pp);
+    return 0;
+}
+
+/* returns 0 for equal, -1 for a < b, 1 for a > b, <= -2 on error */
+static int X509_PUBKEY_cmp(X509_PUBKEY *a, X509_PUBKEY *b)
+{
+    X509_ALGOR *algA = NULL, *algB = NULL;
+    int res = 0;
+
+    if (a == b)
+        return 0;
+    if (a == NULL)
+        return -1;
+    if (b == NULL)
+        return 1;
+    if (!X509_PUBKEY_get0_param(NULL, NULL, NULL, &algA, a)
+        ||
+        !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b))
+        return -2;
+    if (algA == NULL || algB == NULL)
+        return -3;
+    if ((res = X509_ALGOR_cmp(algA, algB)) != 0)
+        return res;
+    return EVP_PKEY_cmp(X509_PUBKEY_get0(a), X509_PUBKEY_get0(b));
+}
+
+/* verifies the Proof-of-Possession of the request with the given rid in reqs */
+int OSSL_CRMF_MSGS_verify_popo(const OSSL_CRMF_MSGS *reqs,
+                               int rid, int acceptRAVerified)
+{
+    OSSL_CRMF_MSG *req = NULL;
+    X509_PUBKEY *pubkey = NULL;
+    OSSL_CRMF_POPOSIGNINGKEY *sig = NULL;
+
+    if (reqs == NULL
+        || (req = sk_OSSL_CRMF_MSG_value(reqs, rid)) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSGS_VERIFY_POPO,
+                CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    switch (req->popo->type) {
```

Comment:
Can `req->popo` ever by NULL?

Reply:
probably it could...

Reply:
Indeed it could, since the field is OPTIONAL. I've added a NULL check.